### PR TITLE
[vm_topology.py]: Vlan interface addition delay handling

### PR DESCRIPTION
**Pull request by praveen-li**
Vlan interface addition may take few secs to reflect in OVS Command, which causes entire add-topo task to fail. It is good to have more concrete fail check. As of now added a code to Retry and Exception.

Signed-off-by: Praveen Chaudhary <pchaudhary@linkedin.com>

### Description of PR
Vlan interface addition may take few secs to reflect in OVS Command, which causes entire add-topo task to fail. It is good to have more concrete fail check. As of now added a code to Retry and Exception.

Fixes # (issue)

### Type of change
- [ ] Bug fix
- [-] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Approach
How did you do it?
Added retry part in case of delay. Retries count is 3.
How did you verify/test it?
Part of pipeline:
http://172.25.11.20:8080/user/falco/my-views/view/Sonic_Build_Deploy_And_Test_PipeLine/
Any platform specific information?
Supported testbed topology if it's a new test case?

### CODE REVIEW

File: ansible/roles/vm_set/library/vm_topology.py [hunk 1]
```diff
@@ -405,8 +406,20 @@ def bind_phys_vlan(self, br_name, vlan_iface, injected_iface, vm_iface, disconne
...
+            bindings = VMTopology.get_ovs_port_bindings(br_name)
+            vlan_iface_id = bindings[vlan_iface]
+            if vlan_iface_id is not None:
+                break
+            time.sleep(1)
+            retries += 1
+        if vlan_iface_id is None:
+            raise Exception("Can't find vlan_iface_id")
+
         injected_iface_id = bindings[injected_iface]
```
Comment (pavel-shirshov) [inline ansible/roles/vm_set/library/vm_topology.py]: Can you please hide retry logic inside of get_ovs_port_bindings()?

File: ansible/roles/vm_set/library/vm_topology.py [hunk 2]
```diff
@@ -406,20 +406,10 @@ def bind_phys_vlan(self, br_name, vlan_iface, injected_iface, vm_iface, disconne
...
-                break
-            time.sleep(1)
-            retries += 1
-        if vlan_iface_id is None:
+        bindings, error = VMTopology.get_ovs_port_bindings(br_name, vlan_iface)
+        if error:
             raise Exception("Can't find vlan_iface_id")
-
+        vlan_iface_id = bindings[vlan_iface]
         injected_iface_id = bindings[injected_iface]
```
Comment (praveen-li) [inline ansible/roles/vm_set/library/vm_topology.py]: Gentle Reminder for review.

File: ansible/roles/vm_set/library/vm_topology.py [hunk 3]
```diff
@@ -490,19 +493,33 @@ def get_ovs_br_ports(bridge):
...
-                iface_name = matched.group(2)
-                result[iface_name] = port_id
-
-        return result
+    def get_ovs_port_bindings(bridge, vlan_iface = None):
+        # Vlan interface addition may take few secs to reflect in OVS Command,
+        # Let`s retry few times in that case.
+        retries = 0
+        vlan_iface_id = None
+        while retries < RETRIES:
```
Comment (pavel-shirshov) [inline ansible/roles/vm_set/library/vm_topology.py]: I'd better to use `for retries in xrange(RETRIES):`

File: ansible/roles/vm_set/library/vm_topology.py [hunk 4]
```diff
@@ -490,19 +493,33 @@ def get_ovs_br_ports(bridge):
...
+            for line in lines:
+                matched = re.match(r'^\s+(\S+)\((\S+)\):\s+addr:.+$', line)
+                if matched:
+                    port_id = matched.group(1)
+                    iface_name = matched.group(2)
+                    result[iface_name] = port_id
+            if vlan_iface is None:
+                return result, False
+            # Check if we have vlan_iface populated
+            vlan_iface_id = result[vlan_iface]
```
Comment (pavel-shirshov) [inline ansible/roles/vm_set/library/vm_topology.py]: how can you be sure you have vlan_iface key in the result? I think you need to check that you have vlan_iface key in the result first

File: ansible/roles/vm_set/library/vm_topology.py [hunk 5]
```diff
@@ -490,19 +493,33 @@ def get_ovs_br_ports(bridge):
...
+            if vlan_iface_id:
+                return result, False
+            time.sleep(1)
```
Comment (pavel-shirshov) [inline ansible/roles/vm_set/library/vm_topology.py]: make the sleep time increasing in some way?
1,2,3 seconds
or
1, 3, 5 seconds

File: ansible/roles/vm_set/library/vm_topology.py [hunk 6]
```diff
@@ -405,7 +406,9 @@ def bind_phys_vlan(self, br_name, vlan_iface, injected_iface, vm_iface, disconne
         if vlan_iface not in ports:
             VMTopology.cmd('ovs-vsctl add-port %s %s' % (br_name, vlan_iface))
 
-        bindings = VMTopology.get_ovs_port_bindings(br_name)
+        bindings, error = VMTopology.get_ovs_port_bindings(br_name, vlan_iface)
+        if error:
+            raise Exception("Can't find vlan_iface_id")
```
[the remark this answers is not in the document]
Comment (praveen-li) [reply ansible/roles/vm_set/library/vm_topology.py]: Sure I will change that part.

[the remark this answers is not in the document]
Comment (praveen-li) [reply ansible/roles/vm_set/library/vm_topology.py]: Sure will raise in function get_ovs_port_bindings.

File: ansible/roles/vm_set/library/vm_topology.py (hunk 3 above)
Comment (praveen-li) [reply ansible/roles/vm_set/library/vm_topology.py]: I will change this to range. range takes more space than xrange but it is supported in python3 as well.

File: ansible/roles/vm_set/library/vm_topology.py (hunk 4 above)
Comment (praveen-li) [reply ansible/roles/vm_set/library/vm_topology.py]: Sure let me add a check for this.

File: ansible/roles/vm_set/library/vm_topology.py (hunk 5 above)
Comment (praveen-li) [reply ansible/roles/vm_set/library/vm_topology.py]: i can use 2*retries +1.

**Comment by praveen-li** [comment]
Incorporated Review comments, kindly Review. Thanks a lot.

**Pull request merged**